File: server_integration_tests/java/java_server_reactive/src/main/java/dev/graphlink/javaserver/impl/NestingServiceImpl.java
package dev.graphlink.javaserver.impl;

import dev.graphlink.javaserver.generated.enums.Color;
import dev.graphlink.javaserver.generated.services.NestingService;
import dev.graphlink.javaserver.generated.types.Box;
import dev.graphlink.javaserver.generated.types.Circle;
import dev.graphlink.javaserver.generated.interfaces.Shape;
import dev.graphlink.javaserver.generated.interfaces.Media;
import dev.graphlink.javaserver.generated.types.Photo;
import dev.graphlink.javaserver.generated.types.Square;
import dev.graphlink.javaserver.generated.types.Video;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Flux;

import java.util.List;

/** List-depth (1/2/3) × nullability × kind (type / interface / union / enum) resolvers. */
@Service
public class NestingServiceImpl implements NestingService {

    private Box box(String id) {
        return Box.builder().id(id).label("label-" + id).build();
    }

    private Circle circle(String id) {
        return Circle.builder().id(id).kind("circle").radius(1).build();
    }

    private Square square(String id) {
        return Square.builder().id(id).kind("square").side(2).build();
    }

    private Photo photo(String id) {
        return Photo.builder().id(id).url("https://x/" + id + ".jpg").width(640).build();
    }

    private Video video(String id) {
        return Video.builder().id(id).url("https://x/" + id + ".mp4").durationSec(30).build();
    }

    @Override
    public Flux<Color> colors1() {
        return Flux.fromIterable(List.of(Color.RED, Color.GREEN, Color.BLUE));
    }

    @Override
    public Flux<List<Color>> colors2() {
        return Flux.fromIterable(List.of(List.of(Color.RED), List.of(Color.GREEN, Color.BLUE)));
    }

    @Override
    public Flux<List<? extends List<Color>>> colors3() {
        return Flux.<List<? extends List<Color>>>fromIterable(List.of(List.of(List.of(Color.RED))));
    }

    @Override
    public Flux<Box> boxes1() {
        return Flux.fromIterable(List.of(box("b1"), box("b2")));
    }

    @Override
    public Flux<List<Box>> boxes2() {
        return Flux.fromIterable(List.of(List.of(box("b1")), List.of(box("b2"))));
    }

    @Override
    public Flux<List<? extends List<Box>>> boxes3() {
        return Flux.<List<? extends List<Box>>>fromIterable(List.of(List.of(List.of(box("b1")))));
    }

    @Override
    public Flux<Shape> shapes1() {
        return Flux.<Shape>fromIterable(List.of(circle("c1"), square("s1")));
    }

    @Override
    public Flux<List<? extends Shape>> shapes2() {
        return Flux.<List<? extends Shape>>fromIterable(List.of(List.of(circle("c1")), List.of(square("s1"))));
    }

    @Override
    public Flux<List<? extends List<? extends Shape>>> shapes3() {
        return Flux.<List<? extends List<? extends Shape>>>fromIterable(List.of(List.of(List.of(circle("c1"), square("s1")))));
    }

    @Override
    public Flux<Media> media1() {
        return Flux.<Media>fromIterable(List.of(photo("p1"), video("v1")));
    }

    @Override
    public Flux<List<? extends Media>> media2() {
        return Flux.<List<? extends Media>>fromIterable(List.of(List.of(photo("p1")), List.of(video("v1"))));
    }

    @Override
    public Flux<List<? extends List<? extends Media>>> media3() {
        return Flux.<List<? extends List<? extends Media>>>fromIterable(List.of(List.of(List.of(photo("p1"), video("v1")))));
    }
}

File: server_integration_tests/kotlin/kotlin_server_blocking/src/main/kotlin/dev/graphlink/kotlinserverblocking/impl/NestingServiceImpl.kt
package dev.graphlink.kotlinserverblocking.impl

import dev.graphlink.kotlinserverblocking.generated.enums.Color
import dev.graphlink.kotlinserverblocking.generated.interfaces.Media
import dev.graphlink.kotlinserverblocking.generated.interfaces.Shape
import dev.graphlink.kotlinserverblocking.generated.services.NestingService
import dev.graphlink.kotlinserverblocking.generated.types.Box
import dev.graphlink.kotlinserverblocking.generated.types.Circle
import dev.graphlink.kotlinserverblocking.generated.types.Photo
import dev.graphlink.kotlinserverblocking.generated.types.Square
import dev.graphlink.kotlinserverblocking.generated.types.Video
import org.springframework.stereotype.Service

/** List-depth (1/2/3) × nullability × kind (type / interface / union / enum) resolvers. */
@Service
class NestingServiceImpl : NestingService {

    private fun box(id: String) = Box(id = id, label = "label-$id")
    private fun circle(id: String) = Circle(id = id, kind = "circle", radius = 1)
    private fun square(id: String) = Square(id = id, kind = "square", side = 2)
    private fun photo(id: String) = Photo(id = id, url = "https://x/$id.jpg", width = 640)
    private fun video(id: String) = Video(id = id, url = "https://x/$id.mp4", durationSec = 30)

    override fun colors1(): List<Color> = listOf(Color.RED, Color.GREEN, Color.BLUE)

    override fun colors2(): List<List<Color?>?>? = listOf(listOf(Color.RED), listOf(Color.GREEN, Color.BLUE))

    override fun colors3(): List<List<List<Color>>> = listOf(listOf(listOf(Color.RED)))

    override fun boxes1(): List<Box?>? = listOf(box("b1"), box("b2"))

    override fun boxes2(): List<List<Box>> = listOf(listOf(box("b1")), listOf(box("b2")))

    override fun boxes3(): List<List<List<Box?>?>?>? = listOf(listOf(listOf(box("b1"))))

    override fun shapes1(): List<Shape> = listOf(circle("c1"), square("s1"))

    override fun shapes2(): List<List<Shape?>?>? = listOf(listOf(circle("c1")), listOf(square("s1")))

    override fun shapes3(): List<List<List<Shape>>> = listOf(listOf(listOf(circle("c1"), square("s1"))))

    override fun media1(): List<Media> = listOf(photo("p1"), video("v1"))

    override fun media2(): List<List<Media?>?>? = listOf(listOf(photo("p1")), listOf(video("v1")))

    override fun media3(): List<List<List<Media>>> = listOf(listOf(listOf(photo("p1"), video("v1"))))
}
